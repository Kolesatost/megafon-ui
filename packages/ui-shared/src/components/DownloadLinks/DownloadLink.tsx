import * as React from 'react';
import { TextLink } from '@megafon/ui-core';
import { cnCreate, filterDataAttrs } from '@megafon/ui-helpers';
import DownloadIcon from '@megafon/ui-icons/basic-32-download_32.svg';
import * as PropTypes from 'prop-types';
import './DownloadLink.less';

export interface IDownloadLink {
    /** Дополнительные data атрибуты к внутренним элементам */
    dataAttrs?: {
        root?: Record<string, string>;
        link?: Record<string, string>;
    };
    /** Ссылка на корневой элемент */
    rootRef?: React.Ref<HTMLDivElement>;
    /** Адресы ссылки */
    href: string;
    /** Текст ссылки */
    text: string;
    /** Расширение файла */
    extension: string;
    /** Размер файла */
    fileSize: string;
    /** Дополнительный класс для компонента */
    className?: string;
    /** Дополнительные классы для корневого и внутренних элементов */
    classes?: {
        root?: string;
        link?: string;
    };
    /** Обработчик клика по ссылке */
    onClick?: (e: React.SyntheticEvent<EventTarget>) => void;
}

const cn = cnCreate('mfui-download-link');
const DownloadLink: React.FC<IDownloadLink> = ({
    href,
    text,
    extension,
    fileSize,
    onClick,
    className,
    classes = {},
    rootRef,
    dataAttrs,
}) => (
    <div {...filterDataAttrs(dataAttrs?.root)} className={cn([className, classes.root])} ref={rootRef}>
        <div className={cn('icon')}>
            <DownloadIcon className={cn('icon-svg')} />
        </div>
        <div>
            <TextLink
                download
                href={href}
                onClick={onClick}
                dataAttrs={{ root: dataAttrs?.link }}
                className={cn('link', [classes.link])}
            >
                {text}
            </TextLink>
            <p className={cn('info')}>{`${extension}${extension && fileSize ? ',' : ''} ${fileSize}`}</p>
        </div>
    </div>
);

DownloadLink.propTypes = {
    dataAttrs: PropTypes.shape({
        root: PropTypes.objectOf(PropTypes.string.isRequired),
        link: PropTypes.objectOf(PropTypes.string.isRequired),
    }),
    rootRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.oneOfType([PropTypes.shape({ current: PropTypes.elementType }), PropTypes.any]),
    ]),
    href: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    extension: PropTypes.string.isRequired,
    fileSize: PropTypes.string.isRequired,
    className: PropTypes.string,
    classes: PropTypes.shape({
        root: PropTypes.string,
        link: PropTypes.string,
    }),
    onClick: PropTypes.func,
};

export default DownloadLink;
