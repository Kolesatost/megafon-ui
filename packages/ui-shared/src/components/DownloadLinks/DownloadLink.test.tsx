import * as React from 'react';
import { mount, shallow } from 'enzyme';
import DownloadLink, { IDownloadLink } from './DownloadLink';

const props: IDownloadLink = {
    href: 'href',
    text: 'text',
    extension: 'PDF',
    fileSize: '500 MB',
};

describe('DownloadLink', () => {
    it('render component', () => {
        const wrapper = shallow(<DownloadLink {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('render with custom classes and data-attributes', () => {
        const wrapper = shallow(
            <DownloadLink
                {...props}
                className="custom-classname"
                classes={{
                    root: 'root-class',
                    link: 'link-class',
                }}
                dataAttrs={{ root: { 'data-testid': 'root-test' }, link: { 'data-testid': 'link-test' } }}
            />,
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should call onClick props', () => {
        const onClick = jest.fn();
        const wrapper = shallow(<DownloadLink {...props} onClick={onClick} />);
        const link = wrapper.find('TextLink');

        link.simulate('click');
        expect(onClick).toBeCalled();
    });

    it('should return reference to root element', () => {
        const ref: React.RefObject<HTMLDivElement> = React.createRef();
        mount(<DownloadLink {...props} rootRef={ref} />);

        expect(ref.current).not.toBeNull();
    });
});
