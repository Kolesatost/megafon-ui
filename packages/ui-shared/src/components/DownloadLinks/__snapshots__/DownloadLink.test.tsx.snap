// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadLink render component 1`] = `
<div
  className="mfui-download-link"
>
  <div
    className="mfui-download-link__icon"
  >
    <test-file-stub
      className="mfui-download-link__icon-svg"
    />
  </div>
  <div>
    <TextLink
      className="mfui-download-link__link"
      color="blue"
      download={true}
      href="href"
      underlineStyle="solid"
      underlineVisibility="hover"
    >
      text
    </TextLink>
    <p
      className="mfui-download-link__info"
    >
      PDF, 500 MB
    </p>
  </div>
</div>
`;

exports[`DownloadLink render with custom classes 1`] = `
<div
  className="mfui-download-link custom-classname root-class"
>
  <div
    className="mfui-download-link__icon"
  >
    <test-file-stub
      className="mfui-download-link__icon-svg"
    />
  </div>
  <div>
    <TextLink
      className="mfui-download-link__link link-class"
      color="blue"
      download={true}
      href="href"
      underlineStyle="solid"
      underlineVisibility="hover"
    >
      text
    </TextLink>
    <p
      className="mfui-download-link__info"
    >
      PDF, 500 MB
    </p>
  </div>
</div>
`;
